### PR TITLE
Invoke via endpoints lister

### PR DESCRIPTION
**Pull request by alexellis**
## Description

Invoke via endpoints lister

## Motivation and Context

- [x] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))

Fixes: #102 

This patch makes the operator use endpoint load-balancing using
a random distribution and is a copy of the approach used in
faas-netes. It is used on an opt-in basis depending on whether the
gateway is set to `direct_functions=false`. 

## How Has This Been Tested?
    
At present, it fails with "error listing" despite updating local

Simply:

* Deploy the new version of the operator:

```
helm repo update && helm upgrade openfaas --install openfaas/openfaas    \
 --namespace openfaas    \
  --set basic_auth=true  \
   --set functionNamespace=openfaas-fn --set clusterRole=false \
  --set operator.create=true --set gateway.directFunctions=false \
  --set gateway.image=openfaas/gateway:0.18.3 \ 
  --set operator.image=openfaas/openfaas-operator:0.12.1-rc5
```

* Set `openfaas/openfaas-operator:0.12.1-rc5` with your own image

* Invoke the function

* See the error
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Impact to existing users

Disable `direct_functions` in the gateway.

Sample image for testing: `openfaas/openfaas-operator:0.12.1-rc1`

RBAC permissions also need to be edited to include endpoints.

Edit the role: `kubectl edit -n openfaas-fn role/openfaas-operator-rw`

Add as per:

```yaml
- apiGroups:
  - ""
  resources:
  - pods
  - pods/log
  - namespaces
  - endpoints
  verbs:
  - get
  - list
  - watch
```

### CODE REVIEW

**Comment by alexellis** [comment]
The error I'm seeing at present is:

```
2019/11/25 17:22:20 resolver error: cannot find figlet: error listing figlet.openfaas-fn endpoints "figlet" not found
2019/11/25 17:22:22 resolver error: cannot find nodeinfo: error listing nodeinfo.openfaas-fn endpoints "nodeinfo" not found
```

**Comment by alexellis** [comment]
The endpoint(s) do exist:
```
kubectl get endpoints -n openfaas-fn figlet
NAME     ENDPOINTS         AGE
figlet   10.32.0.36:8080   166m
```

**Comment by alexellis** [comment]
```
helm upgrade openfaas --install openfaas/openfaas     --namespace openfaas      --set basic_auth=true     --set functionNamespace=openfaas-fn --set clusterRole=false  --set operator.create=true --set operator.image=openfaas/openfaas-operator:0.12.1-rc1 --set gateway.directFunctions=false
```

**Comment by alexellis** [comment]
Still no luck:

```sh
alex@nuc7:~$ kubectl logs -n openfaas deploy/gateway operator -f
I1125 18:25:42.559223       1 main.go:50] Starting OpenFaaS controller version: latest-dev commit: 68e6b7045e170a89e8388c0e5ced940e101d62c8
W1125 18:25:42.559478       1 client_config.go:548] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
I1125 18:25:42.563661       1 controller.go:120] Setting up event handlers
I1125 18:25:42.563741       1 controller.go:173] Waiting for informer caches to sync
2019/11/25 18:25:42 Default ns: "openfaas-fn"
I1125 18:25:42.565559       1 server.go:104] Using namespace 'openfaas-fn'
I1125 18:25:42.565592       1 server.go:105] Starting HTTP server on port 8081
I1125 18:25:42.663970       1 controller.go:178] Starting workers
I1125 18:25:42.664026       1 controller.go:184] Started workers
I1125 18:25:51.628254       1 apply.go:71] Function nodeinfo created
I1125 18:25:51.634465       1 controller.go:273] Creating deployment for 'nodeinfo'
I1125 18:25:51.656034       1 controller.go:285] Creating ClusterIP service for 'nodeinfo'
2019/11/25 18:25:58 Lister for: "openfaas-fn"
2019/11/25 18:25:58 resolver error: cannot find nodeinfo: error listing "nodeinfo"."openfaas-fn", error: "endpoints \"nodeinfo\" not found"
2019/11/25 18:26:05 resolver error: cannot find nodeinfo: error listing "nodeinfo"."openfaas-fn", error: "endpoints \"nodeinfo\" not found"
2019/11/25 18:27:28 resolver error: cannot find nodeinfo: error listing "nodeinfo"."openfaas-fn", error: "endpoints \"nodeinfo\" not found"
^C
alex@nuc7:~$ kubectl get endpoints -n openfaas-fn
NAME       ENDPOINTS         AGE
nodeinfo   10.32.0.12:8080   110s
alex@nuc7:~$ kubectl version
Client Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.3", GitCommit:"b3cbbae08ec52a7fc73d334838e18d17e8512749", GitTreeState:"clean", BuildDate:"2019-11-13T11:23:11Z", GoVersion:"go1.12.12", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.3", GitCommit:"b3cbbae08ec52a7fc73d334838e18d17e8512749", GitTreeState:"clean", BuildDate:"2019-11-13T11:13:49Z", GoVersion:"go1.12.12", Compiler:"gc", Platform:"linux/amd64"}
alex@nuc7:~$ 
```

**Comment by alexellis** [comment]
Latest progress -> the informer appears to be empty.

Screenshot 2019-11-26 at 09 59 45

File: main.go [hunk 1]
```diff
@@ -119,10 +127,12 @@ func main() {
 		factory,
 	)
 
-	go kubeInformerFactory.Start(stopCh)
+	srv := server.New(faasClient, kubeClient, endpointsInformer, deploymentInformer)
```
Comment (LucasRoesler) [inline main.go]: if `srv` is not referenced anywhere else, why move it like this? do you have something planned?

Comment (alexellis) [reply main.go]: This was a suggestion by @munnerz to separate the serving from the initialisation.

File: main.go [hunk 2]
```diff
@@ -119,10 +127,12 @@ func main() {
...
+
 	go faasInformerFactory.Start(stopCh)
-	go server.Start(faasClient, kubeClient, kubeInformerFactory)
+	go kubeInformerFactory.Start(stopCh)
 
+	go srv.Start()
 	if err = ctrl.Run(1, stopCh); err != nil {
```
Comment (alexellis) [inline main.go]: @LucasRoesler it is referenced here, not sure if you saw that?

**Comment by alexellis** [comment]
```
kubectl logs deploy/gateway -n openfaas -f operator
I1126 11:10:55.284061       1 main.go:51] Starting OpenFaaS controller version: latest-dev commit: fe5d5c8c4fc960d4ef3ab1a5ee312f8f74af2a14
W1126 11:10:55.284218       1 client_config.go:548] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
2019/11/26 11:10:55 Waiting for cache sync in main
2019/11/26 11:10:55 Cache sync done
I1126 11:10:55.285847       1 controller.go:120] Setting up event handlers
I1126 11:10:55.285961       1 server.go:104] Using namespace 'openfaas-fn'
I1126 11:10:55.285999       1 controller.go:173] Waiting for informer caches to sync
I1126 11:10:55.286101       1 server.go:119] Starting HTTP server on port 8081
I1126 11:10:55.386202       1 controller.go:178] Starting workers
I1126 11:10:55.386297       1 controller.go:184] Started workers
2019/11/26 11:11:14 nodeinfo took 0.133002 seconds
2019/11/26 11:11:17 figlet took 0.004929 seconds
2019/11/26 11:14:07 figlet took 0.004601 seconds
2019/11/26 11:14:13 figlet took 0.005010 seconds
```

This is what I have now with Kubernetes 1.16 ^

**Comment by alexellis** [comment]
James suggested creating the informers in the root of the application and passing them down, which appeared to get us past the error I was encountering.

**Pull request merged**